anndata/base.py

```diff
@@ -1561,6 +1513,123 @@ def to_df(self) -> pd.DataFrame:
+        if use_raw and is_layer:
+            raise ValueError(
+                "Cannot use expression from both layer and raw. You provided:"
+                "'use_raw={}' and 'layer={}'".format(use_raw, layer)
+            )
+        if is_layer:
+            return self.layers[layer]
+        elif use_raw:
+            if self.raw is None:
+                raise ValueError("This AnnData doesn't have a value in `.raw`.")
+            return self.raw.X
+        else:
+            return self.X
+
+    def obs_array(
+        self, k: str, *, use_raw: bool = False, layer: Optional[str] = None
+    ) -> np.ndarray:
+        """
+        Convenience function for returning a 1 dimensional ndarray of values
+        from `.X`, `.raw.X`, `.layers[k]`, or `.obs`.
+
+        Made for convenience, not performance. Intentionally permissive about
+        arguments, for easy iterative use.
+
+        Params
+        ------
+        k
+            Key to use. Should be in `.var_names` or `.obs.columns`. If `use_raw`,
+            value should be in `.raw.var_names` instead of `.var_names`.
+        use_raw
+            Whether values should be returned from `.raw.X`
+        layer
+            What layer values should be returned from. If `None`, `.X` is used.
+
+        Returns
+        -------
+        A one dimensional nd array, with values for each obs in the same order
+        as `.obs_names`.
+        """
+        if layer == "X":
+            if "X" in self.layers:
+                pass
+            else:
+                # TODO: Raise deprecation warning
+                layer = None
+
+        if k in self.obs:
+            a = self.obs[k].values
+        elif use_raw:
+            a = self.raw[:, k].X
+        else:
+            idx = self._normalize_indices((slice(None), k))
+            a = self._get_expression(layer=layer)[idx]
+        if issparse(a):
+            a = a.toarray().reshape(-1)
+        return a
+
+
+    def var_array(
+        self, k, *, use_raw: bool = False, layer: Optional[str] = None
+    ) -> np.ndarray:
+        """
+        Convenience function for returning a 1 dimensional ndarray of values
+        from `.X`, `.raw.X`, `.layers[k]`, or `.obs`.
+
+        Made for convenience, not performance. Intentionally permissive about
+        arguments, for easy iterative use.
+
+        Params
+        ------
+        k
+            Key to use. Should be in `.obs_names` or `.var.columns`.
+        use_raw
+            Whether values should be returned from `.raw.X`
+        layer
+            What layer values should be returned from. If `None`, `.X` is used.
+
+        Returns
+        -------
+        A one dimensional nd array, with values for each var in the same order
+        as `.var_names`.
+        """
+        if layer == "X":
+            if "X" in self.layers:
+                pass
+            else:
+                # TODO: Raise deprecation warning
+                layer = None
```

Comment:
I just checked the previous implementations of this, which got really complicated by several people iterating on it... This here is much better.

I would do:
```
if use_raw:
     obj = self.raw
else:
     obj = self
```
So that if `use_raw==True` this looks into `self.raw.var`.

Reply:
Hmm. I don't think I like that the `var` accessor could return something with a different shape that `adata.shape[1]`. What if you have to ask `adata.raw` for those values? Like `adata.raw.var_array`?

Reply:
Your alternative solution is 100% fine, too!